File: hphp/hack/src/decl/decl_subst.ml
(*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)

open Hh_prelude
open Typing_defs
module Reason = Typing_reason

type decl_subst = decl_ty SMap.t

(*****************************************************************************)
(* Builds a substitution out of a list of type parameters and a list of types.
 *
 * Typical use-case:
 *   class Y<T> { ... }
 *   class X extends Y<int>
 *
 * To build the type of X, we need to replace all the occurrences of T in Y by
 * int. The function make_subst, builds the substitution (the map associating
 * types to a type parameter name), in this case, it would build the map(T =>
 * int).
 *)
(*****************************************************************************)

let make_locl tparams tyl =
  (* We tolerate missing types in silent_mode. When that happens, we bind
   * all the parameters we can, and bind the remaining ones to "Tany".
   *)
  let make_subst_tparam (subst, tyl) t =
    let (ty, tyl) =
      match tyl with
      | [] ->
        ((Typing_defs.mk (Reason.Rnone, Typing_defs.make_tany ()) : locl_ty), [])
      | ty :: rl -> (ty, rl)
    in
    (SMap.add (snd t.tp_name) ty subst, tyl)
  in
  let (subst, _) =
    List.fold tparams ~init:(SMap.empty, tyl) ~f:make_subst_tparam
  in
  subst

let make_decl tparams tyl =
  (* We tolerate missing types in silent_mode. When that happens, we bind
   * all the parameters we can, and bind the remaining ones to "Tany".
   *)
  let make_subst_tparam (subst, tyl) t =
    let (ty, tyl) =
      match tyl with
      | [] ->
        ((Typing_defs.mk (Reason.Rnone, Typing_defs.make_tany ()) : decl_ty), [])
      | ty :: rl -> (ty, rl)
    in
    (SMap.add (snd t.tp_name) ty subst, tyl)
  in
  let (subst, _) =
    List.fold tparams ~init:(SMap.empty, tyl) ~f:make_subst_tparam
  in
  subst

File: hphp/hack/src/decl/decl_subst.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.
use oxidized_by_ref::typing_defs_core::Tparam;
use typing_collections_rust::SMap;
use typing_defs_rust::typing_defs_core::Ty;
use typing_defs_rust::typing_make_type::TypeBuilder;

pub fn make_locl<'a>(
    bld: &'a TypeBuilder<'a>,
    tparams: impl IntoIterator<Item = &'a Tparam<'a>>,
    mut targs: impl Iterator<Item = Ty<'a>>,
) -> SMap<'a, Ty<'a>> {
    // TODO(hrust)
    let mut substs = SMap::empty();
    for tparam in tparams {
        let targ_ty: Ty = match targs.next() {
            Some(ty) => ty,
            None => bld.any(bld.mk_rnone()),
        };
        substs = substs.add(bld, &tparam.name.1, targ_ty)
    }
    substs
}
